engine/src/logger/instance.cpp

```diff
@@ -0,0 +1,288 @@
+#include <logger/instance.hpp>
+#include <algorithm>
+#include <condition_variable>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+#if defined(_WIN32)
+	#include <Windows.h>
+#endif
+
+namespace gen::logger {
+namespace {
+namespace fs = std::filesystem;
+
+void append_timestamp(std::string& out, Clock::time_point const& timestamp, Timestamp const mode) {
+	static auto s_mutex{std::mutex{}};
+	static constexpr std::size_t buf_size_v{64};
+	auto buffer = std::array<char, buf_size_v>{};
+	auto const time = Clock::to_time_t(timestamp);
+	// std::localtime / std::gmtime are not thread safe. cppref:
+	//  pointer to a static internal std::tm object on success, or null pointer otherwise.
+	// The structure may be shared between std::gmtime, std::localtime, and std::ctime,
+	// and may be overwritten on each invocation.
+	auto lock = std::unique_lock{s_mutex};
+	// NOLINTNEXTLINE
+	auto const tm_struct = mode == Timestamp::eUtc ? *std::gmtime(&time) : *std::localtime(&time);
```

Comment:
We have access to the date library through ext, which is the precursor to C++20's Chrono. All of the timezone and date functions there are fully thread-safe, and in instances like this, we should probably use that instead of dealing with the can of worms that std::gmtime and such open.

Here is the documentation to give you an idea of how it works as it's slightly different from how the standard implements it.

https://github.com/HowardHinnant/date/tree/master#readme

Reply:
Gotcha, will take a look. I'd just gone off what exists in `log.cpp` for this.

Reply:
@Rinzii where's the date lib? I can't seem to find it in `ext`, either subdirs or CMake. 😄

Reply:
Ah, it was being added by FetchContent per suggestions from the lib owner, but it seems it got lost in the move to the new repo.

Just add this below to ext and the link is date::date

`
     include( FetchContent )
     FetchContent_Declare( date_src
       GIT_REPOSITORY https://github.com/HowardHinnant/date.git
       GIT_TAG        v3.0.1  # adjust tag/branch/commit as needed
     )
     FetchContent_MakeAvailable(date_src)
     ...
     target_link_libraries (my_target PRIVATE date::date)
`

Reply:
No problem! However, if it's not part of the project yet, I'm gonna argue that it's out of scope of this PR. Happy to create a new Issue tracking that enhancement, and covering in subsequent PRs.

Reply:
That would be acceptable ^^